lib/using_restricted_array.rb

```diff
@@ -6,56 +6,123 @@
 
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def length(array)
-  raise NotImplementedError
+  length = 0
+  index = 0
+  while array[index] != nil
+    length += 1
+    index += 1
+  end
+  return length
 end
 
 # Prints each integer values in the array
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def print_array(array)
-  raise NotImplementedError
+  (length(array)).times do |i|
+    p "#{array[i]} "
+  end
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def search(array, length, value_to_find)
-  raise NotImplementedError
+  length.times do |i|
+    if array[i] == value_to_find
+      return true
+    end
+  end
+  return false
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def find_largest(array, length)
-  raise NotImplementedError
+  return nil if length == 0
+  max = array[0]
+  length.times do |i|
+    if array[i] > max
+      max = array[i]
+    end
+  end
+  return max
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def find_smallest(array, length)
-  raise NotImplementedError
+  return nil if length == 0
+  min = array[0]
+  length.times do |i|
+    if array[i] < min
+      min = array[i]
+    end
+  end
+  return min
 end
 
 # Reverses the values in the integer array in place
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse(array, length)
-  raise NotImplementedError
+  (length/2).times do |i|
+    last = array[length-(i+1)]
+    first = array[i]
+    array[i] = last
+    array[length-(i+1)] = first
+  end
+  return array
 end
 
 # For an array sorted in ascending order, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(log(n))
+# Space complexity: Calculating only for number of new arrays created - O(log(n)), unclear on how method stacks work in memory
 def binary_search(array, length, value_to_find)
```

Comment:
Since you're essentially creating a new array with each iteration, this is actually O(n * log n) time complexity and the same for space complexity I believe.

An interesting recursive solution, have you considered adjusting the header to be:

```suggestion
# Time complexity: O(log(n))
# Space complexity: Calculating only for number of new arrays created - O(log(n)), unclear on how method stacks work in memory
def binary_search(array, length, value_to_find, low = 0, high = length - 1)
```

Then not copying the array with each iteration and just adjusting the high and low bounds.